### ISSUE TITLE
BoringSSL's SSL_CREDENTIAL API

### ISSUE THREAD

**Issue by davidben** [envoyproxy/envoy#32793]
*Title*: BoringSSL's `SSL_CREDENTIAL` API

*Description*:
This is a heaps up that BoringSSL just landed a new API that might be useful for Envoy. Not a request to change anything, just a heads up, and also a request for thoughts if you all would prefer it work differently. I wasn't sure where to send this to feel free to redirect this elsewhere.

For a while, BoringSSL did not have a built-in way to select between RSA and ECDSA certificates. We left that to the caller. However, RSA vs ECDSA selection is [quite complex](https://boringssl.googlesource.com/boringssl/+/5a3faaa2d50b2540c6973531841723f633f388cd/ssl/test/runner/runner.go#19669), particularly at TLS 1.2 (TLS 1.3 dramatically cleans this up). Even within ECDSA, one may also have different curves. (Really ECDSA-P-256 and ECDSA-P-384 should be thought of as completely different key types. Though we continue to recommend sticking with P-256.) And, in the future, there will be post-quantum key types to juggle.

There are also other certificate negotiation decisions to make such as:
* Different kinds of credentials altogether ([delegate credentials](https://www.rfc-editor.org/rfc/rfc9345.html), [raw public keys](https://www.rfc-editor.org/rfc/rfc7250.html), [external PSKs](https://www.rfc-editor.org/rfc/rfc9258.html), perhaps even more [future exciting things](https://davidben.github.io/merkle-tree-certs/draft-davidben-tls-merkle-tree-certs.html))
* We're very eager to have some kind of [negotiation for trust anchors](https://github.com/davidben/tls-trust-expressions/blob/main/explainer.md) to help with PQ transitions and PKI agility

With all those moving parts, we thought it was better to have that in BoringSSL. `SSL_CREDENTIAL` is meant to be the place to configure that. The idea is that we combine everything related to a single "credential" into an object. Credentials may have different types, like X.509 certificate or otherwise. Each credential also has some criteria, based on TLS protocol rules, that determine whether it can be used on a connection. Callers configure an ordered preference list of credentials with BoringSSL, and we'll pick the first one that matches.

This can still be used in tandem with application-specific selection logic like SNI dispatch. The idea is that Envoy would use its criteria to select a _list_ of candidates, such as an ECDSA and RSA certificate for that vhost. It would then configure the candidates, in preference order, with BoringSSL, and BoringSSL will evaluate those according to protocol rules.

*Relevant Links*:
https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#Credentials

**Comment by nezdolik** [comment]
cc @lizan @ggreenway

**Comment by ggreenway** [comment]
Yes, that sounds interesting and useful. Envoy currently has [logic to determine if the client supports ecdsa](https://github.com/envoyproxy/envoy/blob/54bfb489b40ee2b7f14dc86b9be21e2a34ee25ab/source/common/tls/context_impl.cc#L1121). I'd love to have BoringSSL handle that type of thing.

One issue we'd run into is supporting both the old and the new version until there is a FIPS version of BoringSSL with the new Credentials API. We'd have to decide if we want to delay adopting the new API, or `#ifdef` it and have code for both versions.

**Comment by davidben** [comment]
> One issue we'd run into is supporting both the old and the new version until there is a FIPS version of BoringSSL with the new Credentials API. We'd have to decide if we want to delay adopting the new API, or #ifdef it and have code for both versions.

https://github.com/envoyproxy/envoy/issues/38353 may help with this issue.

**Comment by jmcrawford45** [comment]
I’d like to pick this up if there’s still interest in Envoy supporting it. It looks like the original concern from @ggreenway  is resolved for **BoringSSL-FIPS**, but I believe with AWS-LC support we now have a similar problem/issue with no clear path to resolution.

### Proposed shape

Start with a minimal scope by migrating the existing server-side multi-cert to a single `SSL_CTX` with an ordered credential list, keeping Envoy’s SNI dispatch but letting BoringSSL own the RSA/ECDSA/curve selection (per @davidben  framing above). All of this would be behind a reloadable flag so each PR is a no-op until we flip the default.

### Ordering questions

1. **Gating model**  I’m assuming two layers: a build-time capability macro (AWS-LC never compiles the credential path; legacy is its only path) plus the runtime flag for BoringSSL rollout.

2. **Legacy removal** This model treats “flip default on BoringSSL” and “delete the legacy multi‑`SSL_CTX` path” as separate, with the latter deferred indefinitely until AWS-LC reaches parity. In other words, plan to keep both paths coexisting long-term rather than as a short transition. Is that acceptable, or do we want a firmer deprecation stance?

3. **Sequencing** Any preference on PR granularity: land the credential build path first (no selection change), then switch selection in a follow‑up? Or would you prefer a design doc that covers the whole series before any code lands?

**Comment by ggreenway** [comment]
Ignore the lack of support in non-boringssl for now. But we'll certainly need a runtime flag to enable/disable.

I think a brief design doc would be helpful; you can just put it in the comments in this issue. Make sure to consider the various ways certificates are delivered and selected dynamically (ie the on-demand cert selector).
